x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/AuthenticationService.java

```diff
@@ -113,6 +139,86 @@ public AuthenticationService(
             new ApiKeyAuthenticator(apiKeyService, nodeName),
             new RealmsAuthenticator(numInvalidation, lastSuccessfulAuthCache)
         );
+        this.apiKeyService = apiKeyService;
+    }
+
+    public void authenticateRemoteAccess(
+        final String action,
+        final TransportRequest request,
+        final boolean allowAnonymous,
+        final ActionListener<Authentication> authenticationListener
+    ) {
+        if (false == (threadContext.getHeader(AuthenticationField.AUTHENTICATION_KEY) == null)) {
```

Comment:
All this is just a rough draft -- will polish in a stand-alone PR if we're happy with the overall approach.